Coerce networkx node objects to str.

jaydonnell opened the pull request:
Here's an alternative approach to handling custom networkx node objects. It requires the networkx node to implement __str__, but doesn't require the user's networkx code to depend on ipycytoscape.

ianhi commented on ipycytoscape/cytoscape.py:
```diff
@@ -376,15 +376,15 @@ def add_graph_from_networkx(self, g, directed=False):
             node_instance = Node()
             _set_attributes(node_instance, data)
             if 'id' not in data:
-                node_instance.data['id'] = node
+                node_instance.data['id'] = str(node)
```
Can you change this to check if the node is a subclass of `ipycytoscape.Node`? Something like this should work:
```python
        for node, data in g.nodes( data=True):
            if issubclass(type(node), Node):
                node_instance = node
            else:
                node_instance = Node()
                _set_attributes(node_instance, data)
                if 'id' not in data:
                    node_instance.data['id'] = str(node)
```

ianhi commented on ipycytoscape/cytoscape.py:
```diff
@@ -376,15 +376,15 @@ def add_graph_from_networkx(self, g, directed=False):
             edge_instance = Edge()
-            edge_instance.data['source'] = source
-            edge_instance.data['target'] = target
+            edge_instance.data['source'] = str(source)
+            edge_instance.data['target'] = str(target)
```
Similarly here should check if it subclasses `ipycytoscape.Node`.
```suggestion
            edge_instance.data['source'] = source.data['id'] if issubclass(type(source), Node) else str(source)
            edge_instance.data['target'] = target.data['id'] if issubclass(type(target), Node) else str(target)
```

ianhi replied:
@jaydonnell I think implementing this suggestion will help with the problems you are seeing with the edges.

ianhi commented:
Can `edges` be custom classes in networkx? I don't think so, though if they can we should also check for that

MridulS commented:
> Can edges be custom classes in networkx? I don't think so, though if they can we should also check for that

Nope they can't be custom classes (in networkx edges don't live independently like ipycytospace) but there can be attributes which are technically custom "edge" classes.


```
In [3]: class lol_edge:
    ...:     def __init__(self, something):
    ...:         self.something = something
    ...:

In [4]: G.add_edge(1, 2, lol=lol_edge('possible'))

In [5]: G[1][2]['lol'].something
Out[5]: 'possible'
```

jaydonnell commented:
@ianhi Hoping you can give me some guidance here. I updated the PR to check if the node is a subclass of ipycytoscape.Node, and now it renders the nodes correctly, but the edge no longer renders. I added your "test case" to the example notebook where you can see the issue. 

I'm happy to dig more, and anything that will save me some time would be appreciated. What I don't know is how nodes and edges are associated in ipycytoscape, and where to start looking to determine why the sublass'd nodes aren't being mapped to the networkx edge.

ianhi commented:
> What I don't know is how nodes and edges are associated in ipycytoscape

the `id` attribute of the `data` dictionary is the all important thing here. On the typescript side cytoscapejs will take the edge and check that for nodes that have an `id` (as a string) that exactly matches the `source` and `target` attributes of the edge (both taken as strings). In contrast networkx stores a reference to the custom object in the edge so when the code currently calls `str(source)` the resultant string will look something like this: `'<__main__.Node object at 0x7f9634c2ae80>'` instead of what we really want which is `source.data['id']` which is a string that will have been set by the user. So we need to check if either the source or target of an edge is a subclass of `ipycytoscape.Node` and then use the id from that.

jaydonnell commented:
Thanks, I missed that bit the first time. Should be working now.

ianhi commented:
Awesome! Can you do two small things for the example:
1. Add a brief description of the new examples in markdown cells right above
     - Doesn't need to be exhaustive documentation just a short snippet that you think would help someone discovering this for the first time
2. Clear the outputs of the cells in the final commit. With Kernel -> restart and clear all outputs
![image](https://user-images.githubusercontent.com/10111092/88011069-6390a700-cae4-11ea-97fa-de7c44113d5f.png)

jaydonnell commented:
Done. Thanks for helping me through the process.

ianhi commented:
@jaydonnell thanks for the contribution :)

I added a commit to make the kernel associated with notebook and I'll merge after the checks complete again.
